modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/logical/ExposeIndexRule.java

```diff
@@ -71,18 +87,68 @@ public void onMatch(RelOptRuleCall call) {
                 .filter(idx -> filter(igniteTable, idx.indexName(), idx.searchBounds()))
                 .collect(Collectors.toList());
 
+        indexes = applyHints(scan, indexes);
+
         if (indexes.isEmpty()) {
             return;
         }
 
-        Map<RelNode, RelNode> equivMap = new HashMap<>(indexes.size());
+        Map<RelNode, RelNode> equivMap = IgniteUtils.newHashMap(indexes.size());
         for (int i = 1; i < indexes.size(); i++) {
             equivMap.put(indexes.get(i), scan);
         }
 
         call.transformTo(indexes.get(0), equivMap);
     }
 
+    /** Filter actual indexes list and prune-table-scan flag if any index is forced to use. */
+    private static List<IgniteLogicalIndexScan> applyHints(TableScan scan, List<IgniteLogicalIndexScan> indexes) {
+        List<RelHint> hints = HintUtils.hints(scan, INDEX_HINTS);
+
+        if (hints.isEmpty()) {
+            return indexes;
+        }
+
+        Set<String> tblIdxNames = indexes.stream().map(AbstractIndexScan::indexName).collect(Collectors.toSet());
+        Set<String> idxToSkip = new HashSet<>(capacity(tblIdxNames.size()));
+        Set<String> idxToUse = new HashSet<>(capacity(tblIdxNames.size()));
+
+        for (RelHint hint : hints) {
+            Collection<String> hintIdxNames = hint.listOptions;
+            boolean noIndex = hint.hintName.equals(NO_INDEX.name());
+
+            if (hintIdxNames.isEmpty()) {
+                if (noIndex) {
+                    idxToSkip.addAll(CollectionUtils.difference(tblIdxNames, idxToUse));
```

Comment:
This means we can use FORCE_INDEX and NO_INDEX hints together in the same query.
The documentation doesn't says it is valid usage, and contain no example. I guess it is invalid case, otherwise what does the query could mean: 
"SELECT /* FORCE_INDEX('idx'), NOINDEX */
What hit has a priority? I guess this is invalid case.


Also, SQL is declarative language and we shouldn't expect `hints` collection is ordered.

Reply:
The same questions, as from Andrey
```
    public void testSingleTable() throws Exception {
        var sql = "SELECT /*+ NO_INDEX, FORCE_INDEX({}) */ * FROM TBL1 WHERE val2 = 'v' AND val3 = 'v'";
    public void testSingleTable() throws Exception {
        var sql = "SELECT /*+ FORCE_INDEX({}), NO_INDEX */ * FROM TBL1 WHERE val2 = 'v' AND val3 = 'v'";
```
has different behavior for now

Reply:
Fixed

Reply:
@nva seems we need additional tests for such a case, you need to use : assertThrowsSqlException

Reply:
@zstan `ForceIndexHintPlannerTest#testWithMultipleIndexHints` been added for the cases you specified

Reply:
Let's create a ticket to fix examples in documentation.
The examples, which mentioned above, should be explained in documentation.
The example `SELECT /*+ NO_INDEX(TBL1_IDX2), FORCE_INDEX(TBL2_IDX2) */` in "Hint Scope" section should be rewritten.